src/frequenz/sdk/timeseries/grid.py

```diff
@@ -78,7 +81,17 @@ def initialize(
         if grid_connections[0].metadata is None:
             raise RuntimeError("Grid metadata is None")
 
-        fuse = grid_connections[0].metadata.fuse
+        # The component graph does not store the metadata as a
+        # ComponentMetadata object but as a dict and the type-checker believes
+        # it is a ComponentMetadata object.
+        metadata = grid_connections[0].metadata
```

Comment:
This is weird, why is the type wrong in the component graph? I mean why the type-checker believes it is a `ComponentMetadata` object if it's not?

Reply:
The method `components()` in the Component Graph returns a `set[Component]` where each `Component` is made of `component_id: int`, `category: ComponentCategory`, `type: ComponentType` and `metadata: ComponentMetadata`

However a `dict` type for `metadata` field is returned when calling `components()` for the grid component. So it seems to me the type-checker is not working correctly and the implementation in Component Graph might not be right either. Please check the code below for further details

```python
    fuse = Fuse(max_current=Current.from_amperes(123.0))

    components = {
        Component(1, ComponentCategory.GRID, None, GridMetadata(fuse)),
        Component(2, ComponentCategory.METER),
    }
    connections = {
        Connection(1, 2),
    }

    # pylint: disable=protected-access
    graph = gr._MicrogridComponentGraph(components=components, connections=connections)

    components = graph.components()
    assert len(components) == 2
    for component in components:
        if component.component_id == 1:
            assert component.category == ComponentCategory.GRID
            assert component.metadata is not None
            assert isinstance(component.metadata, dict)  # PASSED
            assert isinstance(component.metadata, GridMetadata)  # FAILED
```

Moreover, this was kind of considered expected behaviour according to existent component graph tests, see https://github.com/frequenz-floss/frequenz-sdk-python/blob/v1.x.x/tests/microgrid/test_graph.py#L836-L855

Reply:
OK, I see this problem is well beyond this PR.

Reply:
Created the ticket to fix it https://github.com/frequenz-floss/frequenz-sdk-python/issues/789